main/src/library/Option.flix

```diff
@@ -0,0 +1,201 @@
+    ///
+    def exists[a](f: a -> Bool, o: Option[a]): Bool = getWithDefault(map(f, o), false)
+
+    ///
+    /// Returns `true` if `o` is `Some(v)` and the predicate `f(v)` evaluates to `true` or if `o` is `None`.
+    /// Otherwise returns `false`.
+    ///
+    def forall[a](f: a -> Bool, o: Option[a]): Bool = getWithDefault(map(f, o), true)
+
+    ///
+    /// Returns `o` if `o` is `Some(v)` and the predicate `f(v)` is true. Otherwise returns `None`.
+    ///
+    def filter[a](f: a -> Bool, o: Option[a]): Option[a] = match o with {
+        case None => None
+        case Some(v) => if (f(v)) Some(v) else None
+    }
+
+    ///
+    /// Returns `Some(f(v))` if `o` is `Some(v)`. Otherwise returns `None`.
+    ///
+    def map[a,b](f: a -> b, o: Option[a]): Option[b] = match o with {
+        case None => None
+        case Some(v) => Some(f(v))
+    }
+
+    ///
+    /// Returns `Some(f(v1, v2))` if `o1` is `Some(v1)` and `o2` is `Some(v2)`. Otherwise returns `None`.
+    ///
+    def map2[a,b,c](f: (a, b) -> c, o1: Option[a], o2: Option[b]): Option[c] = match (o1, o2) with {
+        case (None, _) => None
+        case (_, None) => None
+        case (Some(v1), Some(v2)) => Some(f(v1, v2))
+    }
+
+    ///
+    /// Returns `f(v)` if `o` is `Some(v)`. Otherwise returns `None`.
+    ///
+    def flatMap[a,b](f: a -> Option[b], o: Option[a]): Option[b] = match o with {
+        case None => None
+        case Some(v) => f(v)
+    }
+
+    ///
+    /// Returns `f(v1, v2)` if `o1` is `Some(v1)` and `o2` is `Some(v2)`. Otherwise returns `None`.
+    ///
+    def flatMap2[a,b,c](f: (a, b) -> Option[c], o1: Option[a], o2: Option[b]): Option[c] = match (o1, o2) with {
+        case (None, _) => None
+        case (_, None) => None
+        case (Some(v1), Some(v2)) => f(v1, v2)
+    }
+
+    ///
+    /// Returns a one-element list of the value `v` if o is `Some(v)`. Otherwise returns the empty list.
+    ///
+    def toList[a](o: Option[a]): List[a] = match o with {
+        case None => Nil
+        case Some(v) => v :: Nil
+    }
+
+    ///
+    /// Returns a one-element set of the value `v` if `o` is `Some(v)`. Otherwise returns the empty set.
+    ///
+    def toSet[a](o: Option[a]): Set[a] = match o with {
+        case None => #{}
+        case Some(v) => #{v}
+    }
+
+    ///
+    /// Returns `o1` if it is `Some(v)`. Otherwise returns `o2`.
+    ///
+    def withDefault[a](o1: Option[a], o2: Option[a]): Option[a] = match o1 with {
+        case None => o2
+        case Some(v) => o1
+    }
+
+    ///
+    /// Returns `Ok(v)` if `o` is `Some(v)`. Otherwise returns `Ok(d)`.
+    ///
+    def toOk[a,b](o: Option[a], d: a): Result[a,b] = Ok(getWithDefault(o, d))
+
+    ///
+    /// Returns `Err(v)` if `o` is `Some(v)`. Otherwise returns `Err(d)`.
+    ///
+    def toErr[a,b](o: Option[b], d: b): Result[a,b] = Err(getWithDefault(o, d))
```

Comment:
In the same wain, this should be:

```
def toErr[t, e](o: Option[e], e:t): Result[t, e]
```
which maps Some(v) to Err(v) otherwise Ok(t).

Notice how the type parameters are swapped in the signature, e.g. Option[e].